core/domain/classifier_services.py

```diff
@@ -0,0 +1,338 @@
+        probs_i = self._c_dl[d] + (self._b_dl[d] * self._alpha)
+        probs_i = probs_i / probs_i.sum(axis=0)[numpy.newaxis]
+        return probs_i
+
+    def _get_probability_with_label(self, d):
+        """Returns a dict that maps human-readable labels to the probability
+        that the doc has the label."""
+        probs_i = self._get_probability(d)
+        probs_l = {}
+        for label, l in self._label_to_id.iteritems():
+            probs_l[label] = probs_i[l]
+        return probs_l
+
+    def _predict(self, d, threshold=0.5):
+        """Calculates prediction data for a doc.
+        Returns a tuple of (predicted label, prediction score)."""
+        probs_l = self._get_probability_with_label(d)
+        default_prob = probs_l[self._DEFAULT_LABEL]
+        default_prob_comp = 1.0 - default_prob
+        for label in sorted(probs_l, key=probs_l.get, reverse=True):
+            prob_l = probs_l[label]
+            if (label != self._DEFAULT_LABEL and
+                    prob_l / default_prob_comp > threshold):
+                return label, prob_l / default_prob_comp
+        return self._DEFAULT_LABEL, default_prob / default_prob_comp
+
+    def _validate_label(self, label):
+        """Checks that labels don't begin with underscore."""
+        if label[0] == '_':
+            raise Exception('Label %s cannot begin with underscore.' % label)
+
+    def _parse_examples(self, examples):
+        """Splits examples into docs (split on spaces) and labels.
+        Example format:
+        [
+            ['doc1', ['label1']],
+            ['doc2', ['label2']],
+            etc
+        ]
+        """
+        docs = []
+        labels = []
+        for example in examples:
+            doc = example[0].split()
+            if len(doc) > 0:
+                docs.append(doc)
+                map(self._validate_label, example[1])
+                labels.append(example[1])
+        return docs, labels
+
+    def _train_docs(self, iterations, doc_ids=None):
+        """Trains given docs (default all) for a number of iterations."""
+        doc_ids = self._get_doc_ids(doc_ids)
+
+        for i in xrange(iterations):
+            statez = self._infer_docs(doc_ids)
+
+    def load_examples(self, examples, iterations=25):
+        """Sets new examples. Overwrites existing ones."""
+        docs, labels_list = self._parse_examples(examples)
+
+        label_set = set(
+            [self._DEFAULT_LABEL] +
+            [label for labels in labels_list for label in labels])
+
+        self._num_labels = len(label_set)
+        self._label_to_id = dict(zip(label_set, xrange(self._num_labels)))
+
+        self._num_words = 0
+        self._word_to_id = {}
+
+        self._num_docs = len(docs)
+
+        self._b_dl = numpy.array(
+            map(self._get_label_vector, labels_list), dtype=int)
+        self._w_dp = [map(self._get_word_id, doc) for doc in docs]
+        self._l_dp = []
```

Comment:
So -- self._init_docs() is basically updating self._l_dp. Maybe bring this down and put this just before self._init_docs() (or whatever it gets renamed to).

(I still feel that something's not quite right about separating out init_docs() like this, since it breaks parity a bit with the other updates going on in this method -- but I don't have a good idea for what to do, since it looks like that's the only behaviour both methods have in common. Perhaps inspiration will strike on a third pass.)